sdl-core/src/main/scala/io/smartdatalake/workflow/action/MLflowTrainAction.scala

```diff
@@ -0,0 +1,258 @@
+
+/**
+ * @param mlflowId id of the MLflow DataObject
+ * @param runInfoId id of an DataObject that implements the CanCreateDataFrame trait. Here the run information of the training is written down
+ * @param inputIds input DataObjects
+ * @param trainIdSelector id of the DataObject that will be used for training
+ * @param transformers optional list of transformations to apply. See [[spark.transformer]] for a list of included Transformers.
+ *                     The transformations are applied according to the lists ordering.
+ * @param pythonModelCode machine learning model provided directly as python code
+ * @param pythonModelFile machine learning model provided as python file
+ * @param registerModel register model in MLflow registry
+ */
+case class MLflowTrainAction(
+                              override val id: ActionId,
+                              mlflowId: DataObjectId,
+                              runInfoId: DataObjectId,
+                              inputIds: Seq[DataObjectId],
+                              trainIdSelector: Option[String] = None,
+                              transformers: Seq[GenericDfsTransformer] = Seq(),
+                              pythonModelCode: Option[String] = None,
+                              pythonModelFile: Option[String] = None,
+                              registerModel: Boolean = false,
+                              override val breakDataFrameLineage: Boolean = false,
+                              override val persist: Boolean = false,
+                              override val mainInputId: Option[DataObjectId] = None,
+                              override val mainOutputId: Option[DataObjectId] = None,
+                              override val executionMode: Option[ExecutionMode] = None,
+                              override val executionCondition: Option[Condition] = None,
+                              override val metricsFailCondition: Option[String] = None,
+                              override val metadata: Option[ActionMetadata] = None,
+                            )(implicit instanceRegistry: InstanceRegistry)
+  extends DataFrameActionImpl {
+
+  // get model code either from the file are directly from input
+  private val pythonCode = {
+    implicit val defaultHadoopConf: Configuration = new Configuration()
+    pythonModelFile
+      .map(file => HdfsUtil.readHadoopFile(file))
+      .orElse(pythonModelCode)
+      .getOrElse(throw ConfigurationException(s"ML model must either be provided as python file or code"))
+  }
+
+  // handle MLflow in and output data object
+  val mlflow = getInputDataObject[MLflowDataObject](mlflowId)
+  val mlflowRunInfo = getOutputDataObject[DataObject with CanWriteDataFrame with CanCreateDataFrame](runInfoId)
+
+  // handle data in- and outputs
+  override val inputs: Seq[DataObject with CanCreateDataFrame] =
+    inputIds.map(getInputDataObject[DataObject with CanCreateDataFrame]) :+ mlflow

```

Comment:
As we are writing an MLFLow model, i suggest that mlflow DataObject should be an output, not input. But what DataFrame could we write there?
Maybe we should implement CanReceiveScriptNotification instead of CanWriteDataFrame, but then it could no longer extend DataFrameActionImpl.

Reply:
Good catch. The `MLflowtrainAction` does not even necessarily have to implement the `CanWriteparkDataFrame` trait. It could only write a Notification that the training is finished. But somehow this switch between Script and DF is not well defined yet.
Run information is written to `runInfo` DataObject.